core/src/main/java/org/apache/iceberg/IncrAppend.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.io.OutputFile;
+
+
+/**
+ * {@link AppendFiles Append} implementation that adds a new manifest file for the write.
```

Comment:
I don't think this is needed. See `Snapshot#addedFiles`, which can already get the set of files added for any snapshot.

Reply:
`Snapshot#addedFiles` will give me back a list of `DataFiles`. `IncrAppend` is actually being used to generate a staging snapshot, which can then be used for files planning

Reply:
Another problem here could be that all the datafiles can belong to different partition specs, as these files can belong to different manifests. Also, datafiles do not have partition spec info. It is currently unclear to me how to handle that. Maybe a first step is to  only allow incremental processing if all the data files belong to the table's current partition spec?

Reply:
Yeah, I meant that `planFiles` could get all of the added files from appends.

For the different partition specs, we would use the same filtering logic that is currently delegated to ManifestReader and FilteredManifest. Basically, we would just need evaluators to check the each data file's partitions and metrics.